File: Core/src/main/java/me/mrkirby153/kcutils/Module.java
package me.mrkirby153.kcutils;

import org.bukkit.event.Listener;
import org.bukkit.plugin.java.JavaPlugin;

/**
 * Abstract framework for utilities
 */
public abstract class Module<T extends JavaPlugin> {

    private T plugin;
    private String name;
    private String version;
    private boolean loaded = false;

    public Module(String name, String version, T plugin) {
        this.name = name;
        this.plugin = plugin;
        this.version = version;
    }

    /**
     * Gets the name of the module
     *
     * @return The module's name
     */
    public String getName() {
        return name;
    }

    /**
     * Gets the version of the module
     *
     * @return The module's version
     */
    public String getVersion() {
        return version;
    }

    /**
     * Loads the module
     */
    public void load() {
        if (loaded) {
            throw new IllegalArgumentException("Attempting to reload an already loaded module!");
        }
        long startTime = System.currentTimeMillis();
        log("Loading version " + version);
        init();
        long endTime = System.currentTimeMillis();
        long elapsedTime = endTime - startTime;
        log("Loaded in " + (elapsedTime < 1 ? "< 1 millisecond" : Time.format(1, elapsedTime, Time.TimeUnit.FIT)));
        loaded = true;
    }

    /**
     * Gets the plugin that loaded the module
     *
     * @return The plugin
     */
    protected T getPlugin() {
        return plugin;
    }

    protected abstract void init();

    /**
     * Log a message to the plugin's log
     *
     * @param message The message
     */
    protected void log(String message) {
        plugin.getLogger().info(String.format("[%s] %s", name.toUpperCase(), message));
    }

    /**
     * Register a listener
     *
     * @param listener A listener to register
     */
    protected void registerListener(Listener listener) {
        plugin.getServer().getPluginManager().registerEvents(listener, plugin);
    }

    protected void runAsync(Runnable runnable) {
        plugin.getServer().getScheduler().runTaskAsynchronously(plugin, runnable);
    }

    /**
     * Schedules a task to be run later
     *
     * @param runnable The task to be run
     * @param delay    The delay
     */
    protected void runLater(Runnable runnable, long delay) {
        plugin.getServer().getScheduler().runTaskLater(plugin, runnable, delay);
    }

    /**
     * Runs a task later asyncronously
     *
     * @param runnable The task to run
     * @param delay    The delay
     */
    protected void runLaterAsync(Runnable runnable, long delay) {
        plugin.getServer().getScheduler().runTaskLaterAsynchronously(plugin, runnable, delay);
    }

    /**
     * Schedule a repeating task
     *
     * @param runnable     The task to be run
     * @param initialDelay The initial delay of the task
     * @param interval     The interval of which the task should be run
     */
    protected void scheduleRepeating(Runnable runnable, long initialDelay, long interval) {
        plugin.getServer().getScheduler().scheduleSyncRepeatingTask(plugin, runnable, initialDelay, interval);
    }
}

File: Bukkit/src/main/kotlin/me/mrkirby153/kcutils/Module.kt
package me.mrkirby153.kcutils

import org.bukkit.configuration.file.FileConfiguration
import org.bukkit.event.Listener
import org.bukkit.plugin.java.JavaPlugin
import kotlin.system.measureTimeMillis

/**
 * A class for creating "sub-plugins" in one main plugin
 *
 * Modules are used to logically separate parts of plugins from one another.
 *
 * @param name      The name of the plugin
 * @param plugin    The owning [JavaPlugin]
 */
abstract class Module<out T : JavaPlugin>(private val name: String, protected val plugin: T) {

    /**
     * If the module has been loaded
     */
    var loaded: Boolean = false

    /**
     * The configuration file of the main plugin
     */
    val config: FileConfiguration
        get() = plugin.config

    /**
     * Loads the module
     */
    fun load() {
        if (loaded)
            throw IllegalStateException("Attempting to load an already loaded module!")
        val loadTime = measureTimeMillis {
            try {
                init()
            } catch (e: Exception) {
                log("[ERROR] Caught exception while loading. (${e.message})")
                e.printStackTrace()
                return
            }
        }
        loaded = true
        log("Loaded in ${if (loadTime < 1) "< 1 millisecond" else Time.format(1, loadTime, Time.TimeUnit.FIT)}")
    }

    /**
     * Unloads the module
     */
    fun unload() {
        if (!loaded)
            throw IllegalStateException("Attempting to unload an already unloaded module!")
        val unloadTime = measureTimeMillis {
            try {
                disable()
            } catch (e: Exception) {
                log("[ERROR] Caught exception while unloading. (${e.message})")
                e.printStackTrace()
                return
            }
        }
        loaded = false
        log("Loaded in ${if (unloadTime < 1) "< 1 millisecond" else Time.format(1, unloadTime, Time.TimeUnit.FIT)}")
    }

    /**
     * Called when the module is loaded
     */
    open fun init() {

    }

    /**
     * Called when the module is unloaded
     */
    open fun disable() {

    }


    /**
     * Logs a message
     *
     * @param msg The message to log
     */
    fun log(msg: Any) {
        plugin.logger.info(String.format("[%s] %s", name.uppercase(), msg))
    }

    /**
     * Registers a listener
     *
     * @param listener The listener to register
     */
    fun registerListener(listener: Listener) {
        plugin.server.pluginManager.registerEvents(listener, plugin)
    }

    /**
     * Runs a task async
     *
     * @param task The task to run async
     */
    fun runAsync(task: () -> Unit) {
        plugin.server.scheduler.runTaskAsynchronously(plugin, task)
    }

    /**
     * Runs a task async
     *
     * @param runnable The task to run async
     */
    fun runAsync(runnable: Runnable) {
        plugin.server.scheduler.runTaskAsynchronously(plugin, runnable)
    }

    /**
     * Runs a task synchronously
     *
     * @param task The task to run sync
     */
    fun runSync(task: () -> Unit) {
        plugin.server.scheduler.runTask(plugin, task)
    }

    /**
     * Runs a task synchronously
     *
     * @param runnable The task to run sync
     */
    fun runSync(runnable: Runnable) {
        plugin.server.scheduler.runTask(plugin, runnable)
    }


    /**
     * Schedules a task to be run periodically
     *
     * @param delay The initial delay before running the task
     * @param interval The interval in which the task is to run
     * @param task The task to run
     */
    fun scheduleRepeating(delay: Long, interval: Long, task: () -> Unit) {
        plugin.server.scheduler.scheduleSyncRepeatingTask(plugin, task, delay, interval)
    }

    /**
     * Schedules a task to be run periodically
     *
     * @param delay The initial delay before running the task
     * @param interval The interval in which the task is to run
     * @param runnable The task to run
     */
    fun scheduleRepeating(delay: Long, interval: Long, runnable: Runnable) {
        plugin.server.scheduler.scheduleSyncRepeatingTask(plugin, runnable, delay, interval)
    }
}